### PR TITLE
[RFE#1726] refactor: filters: create(Reader|Writer), get(Input|Output)Encoding overrides

### PR DESCRIPTION
**Pull request by miurahr**
Improvement, refactor and style change for AbstractFilter class.

Some filter override `processFile(File, File, fc)`  to enforce encoding, but I prefer to implement `getInputEncoding` and `getOutputEncoding` methods.

## Pull request type

- Other (describe below)
Refactoring

## Which ticket is resolved?

- #1726 AbstractFilter should handle BOM when create reader
- https://sourceforge.net/p/omegat/feature-requests/1726/

## What does this PR change?

- Bump commons-io@2.15.0
    - Modify MagicComment and TMXReader2 classes 
- Use BOMInputStream class where hand-crafted detectors in AbstractFilter class
    - Change fallback default charsets to `StandardCharsets.UTF_8`  rather than `Charset.defaultCharset()`
- Prefer override get(Input|Output)Encoding method instead of create(Reader|Writer)
    - MozillaDTDFilter
    - MozillaLangFilter
    - ResourceBundleFilter
    - MoodlePHPFilter
    - PoFilter 
- Apply spotless
- Fix missing javadoc for AbstractFilter class

### CODE REVIEW

File: src/org/omegat/filters2/text/bundles/ResourceBundleFilter.java [hunk 1]
```diff
@@ -146,50 +142,6 @@ public Instance[] getDefaultInstances() {
...
-     *             If any I/O Error occurs upon reader creation
-     */
-    @Override
-    public BufferedReader createReader(File inFile, String inEncoding) throws IOException {
-        Charset charset;
-        if (inEncoding != null) {
-            charset = Charset.forName(inEncoding);
-        } else {
-            charset = Charset.defaultCharset();
-        }
```
Comment (miurahr) [inline src/org/omegat/filters2/text/bundles/ResourceBundleFilter.java:173]: Here we finally change a buggy behavior. In previous versions, the `ResourceBundleFilter ` can read platform default charset,  such as Windows-1251, that is not compliant with Java standard.

File: src/org/omegat/filters2/po/PoFilter.java [hunk 2]
```diff
@@ -306,51 +317,11 @@ public String getFileFormatName() {
...
-     *            Encoding of the input file, if the filter supports it. Otherwise null.
-     * @return
-     * @throws IOException
-     */
-    @Override
-    protected BufferedReader createReader(File inFile, String inEncoding) throws IOException {
-        BOMInputStream bomInputStream = new BOMInputStream(Files.newInputStream(inFile.toPath()),
-                ByteOrderMark.UTF_8, ByteOrderMark.UTF_16LE);
-        ByteOrderMark bom = bomInputStream.getBOM();
-        String charset;
```
Comment (miurahr) [inline src/org/omegat/filters2/po/PoFilter.java:328]: Now we have `AbstractFilter#createReader` which as same functionality with one here, so we can remove it.

File: src/org/omegat/filters2/AbstractFilter.java [hunk 3]
```diff
@@ -335,13 +326,16 @@ public Map<String, String> changeOptions(Dialog parent, Map<String, String> conf
...
-            charset = Charset.forName(inEncoding);
+        BOMInputStream bomInputStream = BOMInputStream.builder().setFile(inFile)
+                .setByteOrderMarks(ByteOrderMark.UTF_8, ByteOrderMark.UTF_16BE, ByteOrderMark.UTF_16LE).get();
+        bomLastParsedFile = bomInputStream.getBOM();
+        String charset;
+        if (bomLastParsedFile != null) {
+            charset = bomLastParsedFile.getCharsetName();
         } else {
-            charset = Charset.defaultCharset();
+            charset = Objects.requireNonNullElseGet(inEncoding, () -> Charset.defaultCharset().name());
```
Comment (miurahr) [inline src/org/omegat/filters2/AbstractFilter.java:336]:
```suggestion
            charset = Objects.requireNonNullElseGet(inEncoding, () -> StandardCharsets.UTF_8.name());
```

File: src/org/omegat/filters2/AbstractFilter.java [hunk 4]
```diff
@@ -359,8 +354,13 @@ protected BufferedReader createReader(File inFile, String inEncoding)
...
+            return null;
+        }
         Charset charset;
-        if (outEncoding != null) {
+        if (bomLastParsedFile != null) {
+            charset = Charset.forName(bomLastParsedFile.getCharsetName());
+        } else if (outEncoding != null) {
             charset = Charset.forName(outEncoding);
         } else {
             charset = Charset.defaultCharset();
```
Comment (miurahr) [inline src/org/omegat/filters2/AbstractFilter.java:366]:
```suggestion
            charset = StandardCharsets.UTF_8;
```

**Review by miurahr — COMMENTED** [review]
We should not use Default charset such as Windows-1251, Shift-JIS, etc.
We can use `StandardCharsets.UTF_8`

**Comment by miurahr** [comment]
The change here will affected all the filter authors and existing filters.

**Comment by brandelune** [comment]
> The change here will affected all the filter authors and existing filters.

Not only the behavior for the 5 filters you mention in the code description?

**Comment by miurahr** [comment]
> > The change here will affected all the filter authors and existing filters.
> 
> Not only the behavior for the 5 filters you mention in the code description?

AbstractFilter class is able to be used as a base, extends,  to implement IFilter interface that any filter plugin should implement;

**Comment by miurahr** [comment]
All questions are answered.

**Pull request merged**